### PR TITLE
Feat/pandera hook

### PR DESCRIPTION
**Pull request by Galileo-Galilei**
## Description

The goal of this PR is to create a MVP to enable runtime data validation easily in an existing kedro project. 

## Development notes

The goal is to enable the following user workflow in a kedro project: 

1. *Generate a configuration file with the schema of a dataset you want to check at runtime*. I did the following changes: 
- rename ``infer_schema`` to ``infer_dataset_schema`` because of import conflicts with pandera function
- log by default in ``base`` instead of ``local`` because of a bug with ``OmegaConfigLoader`` which cannot resolve the key -> should **open an issue in kedro's repo**. 
- logs in a file ``catalog_{dataset_name}.yaml`` with an extra ``_{dataset_name}`` key to enable discovery in the catalog

2. *Let catalog datasets accept an extra ``metadata.pandera.schema`` key which will be resolved as a pandera ``DataframeSchema``*. I did the following changes:
- add custom resolvers (``pa.dict`` and ``pa.yaml``) which can parse another catalog key and convert it to a ``DataframeSchema``
- add a ``after_context_created`` hook to register these resolvers automatically

3. *Add runtime validation each time a session is run*. I did the following changes:
-  add an ``before_node_run`` which looks for a ``DataframeSchema`` in the dataset (under ``metadata.pandera.schema`` key ) and validate it against data passed to the node. 

I also did the following extra actions: 
- Write a tutorial which describes this workflow on the ``pandas-iris`` starter
- Make default logging to INFO so the plugin logs can be seen by user by default
- extra cleanup for migration to ruff linter 
- Fix readthedocs build

## Checklist

- [x] Read the [contributing](../CONTRIBUTING.md) guidelines
- [X] Open this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Update the documentation to reflect the code changes
- [x] Add a description of this change and add your name to the list of supporting contributions in the [`CHANGELOG.md`](../CHANGELOG.md) file. Please respect [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines.
- [x] Add tests to cover your changes (only partially, to be continued)

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (noklam) [inline kedro_pandera/framework/hooks/pandera_hook.py:48]:
```suggestion
            if catalog._data_sets[name].metadata is not None and if "pandera" in catalog._data_sets[name].metadata:
```
How about this?

File: .readthedocs.yml
Comment (Galileo-Galilei) [reply .readthedocs.yml]: Just a copy paste from the new config file on readthedocs, obviously not needed. 

(BTW the landing page on readthedocs documentation shows a screenshot from some "astrojuanlu" guy, open source world is smaller than we think ;))

File: docs/source/01_introduction/01_tutorial.md
Comment (Galileo-Galilei) [reply docs/source/01_introduction/01_tutorial.md:86]: I've hesitated a lot about this, and don't have strong feelings. 

"kedro-pandera", "kedro_pandera" or "pandera"? Last one if faster to write and I am lazy 😄, but maybe a more explicit key would avoid conflict is someone decided to use this key too. I Guess we should use "kedro_pandera" (with an underscore)  rather than "kedro-pandera" (with a dash) to avoid any potential non valid python name if the dict was converted into a object in the future though.  

Happy to change it. Maybe @datajoely have some thoughts here too?

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (Galileo-Galilei) [reply kedro_pandera/framework/hooks/pandera_hook.py]: Just hallucinated, I though this failed but it works perfectly!

File: docs/source/01_introduction/01_tutorial.md
Comment (Galileo-Galilei) [reply docs/source/01_introduction/01_tutorial.md:20]: I was wondering this too. Honestly  I don't care to commit to retrocompatiblity too soon. I think we could add support for ``TemplatedConfigLoader`` later if it is a strongly required feature and once the design is mature. For my personal experience, apart from the very first versions of the 0.18.X series, you can just upgrade an existing project and everything works fine. I guess people with say >=0.18.3 will be able to upgrade to 0.18.13 and use the plugin with no effort. 

``OmegaConfigLoader`` custom resolvers offer a very neat solution to several problems I had during development (anticipating support for different format of DataFrameSchema like python or yaml, having a very simple syntax to load the schema, and having the schema object directly generated form config without a lot of boilerplate custom loading code inside the hook) and I though the advantages were worth the low retrocompatiblity but it can be challenged later. I guess the goal is to have a mature design for the 0.19.X series, and people will move toward the plugin progressively.  

I'll describe in detail in the PR what I did to have you review before we merge. My sort term goal is rather to find out a way to ensure a smooth developer experience and implement the most desired feature.

File: docs/source/01_introduction/01_tutorial.md
Comment (datajoely) [reply docs/source/01_introduction/01_tutorial.md:86]: I think it's fine like this :)

File: docs/source/01_introduction/01_tutorial.md
Comment (datajoely) [reply docs/source/01_introduction/01_tutorial.md:20]: I'm a fan of not supporting TemplatedConfigLoader - this is a good carrot to get people to update!

Comment (Galileo-Galilei) [reply docs/source/01_introduction/01_tutorial.md:20]: Not sure if you're non partisan here @datajoely but I won't complain since we agree, sorry @noklam ;)
TBH I can't be involved as much as I can (there are 15 opened issues on ``kedro-mlflow`` while I am writing this, kedro-serving is in a stale for 2 years, and I've created the repo for kedro pandera for serveal months and never had time to develop before I was in holidays) so let's not be too ambitious for now. We can obviously refactor it later.

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (datajoely) [inline kedro_pandera/framework/hooks/pandera_hook.py:42]: Super elegant

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (datajoely) [inline kedro_pandera/framework/hooks/pandera_hook.py:55]: I wonder if we need a CLI flag or environment var to disable checks?

Comment (Galileo-Galilei) [reply kedro_pandera/framework/hooks/pandera_hook.py:55]: We sure do, but I think this PR is far too big for a MVP. I think we should ship it with docs/tutorial, and then discuss further on what we want to add. I plan to answer to your comment in #12 too!

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (Galileo-Galilei) [reply kedro_pandera/framework/hooks/pandera_hook.py:42]: The problem is that this function does not exist on any public version of kedro yet, but will be in the next release...

File: docs/source/01_introduction/01_tutorial.md
Comment (noklam) [reply docs/source/01_introduction/01_tutorial.md:86]: @Galileo-Galilei Good point. I am happy with both `kedro_pandera` and `pandera`. I have bias due to https://github.com/kedro-org/kedro-viz/blob/6e1d127fb85349ca7244e76719b756cc5001784f/package/kedro_viz/data_access/repositories/catalog.py#L53-L68.

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (noklam) [reply kedro_pandera/framework/hooks/pandera_hook.py:42]: I think we want to do a release soon - maybe blocked by the `kedro-datasets` Python 3.11 release. This alone is enough for a release IMO.

File: docs/source/01_introduction/01_tutorial.md
Comment (Galileo-Galilei) [reply docs/source/01_introduction/01_tutorial.md]: Let's stay with ``pandera``, ``kedro_pandera`` with an underscore would not really add consistency since ``kedro-viz`` is written with a dash.

File: kedro_pandera/framework/hooks/pandera_hook.py
Comment (Galileo-Galilei) [reply kedro_pandera/framework/hooks/pandera_hook.py:54]: Note: the [environment variable solution seems supported](https://pandera.readthedocs.io/en/stable/pyspark_sql.html#switching-validation-on-and-off), but not sure if it is only for pyspark:

File: docs/source/01_tutorials/01_getting_started.md
Comment (datajoely) [inline docs/source/01_tutorials/01_getting_started.md:34]:
```suggestion
If you are using  ``kedro>=0.18.13, <0.19``, you need to configure your ``settings.py`` to use the (future default) ``OmegaConfigLoader``: which is the only ``ConfigLoader``  compatible with ``kedro-pandera``.
```

**Review by datajoely — APPROVED** [review]
I think this is excellent @Galileo-Galilei , the only question I have if it's better to do a `cookiecutter.vcs.clone` for your `pandas-iris` integration test rather than copying it verbatim into this this repo?

**Comment by noklam** [comment]
@Galileo-Galilei Is there anything particular you would like me to review/test? I see this still in draft, anything I can help to push this to the finishing line?

**Comment by Galileo-Galilei** [comment]
Hi @noklam, I think this is almost good to go. I want to write tests and rebase to make commit more readable but it you want I can merge tintoday without all test so you can work in par allel without  too many merge conflicts to resolve. 

I think the main next step is to support ``DataframeModel`` written in python and not only yaml configuration which is very restrictive. 
Do you have another feature in mind?

**Comment by datajoely** [comment]
FYI - Our teams that have adopted Pandera internally have leveraged the Python approach on the functions called by the nodes. In this situation Kedro isn't really aware of this, but our plugin could introspect the function bounded to the `node` object and bring these into scope.

**Comment by datajoely** [comment]
Additionally - shall we make a little feature backlog, because we can split up the work?

**Comment by noklam** [comment]
@Galileo-Galilei That's fine. This is all I need, I can start working on that tomorrow, I will branch off from this PR. I don't expect major changes so I can relies on existing structure.

@datajoely is that the @check_types annotation?

**Comment by datajoely** [comment]
@noklam yes exactly, the pattern in use looks something like this:

```
project/
├─ data_processing/
│  ├─ nodes.py  <--- Functions called call schemas via type hints
│  ├─ pipelines.py
├─ data_science/
│  ├─ ...
├─ schemas/
│  ├─ raw_preprocessing.py
│  |─ prm_features.py
```

Outside of the node definition, where the bound function is defined the signature looks something like this:

```python
import pandera as pa
from pandera.typing import DataFrame

@pa.check_types
def preprocess_data(raw_data: DataFrame,...) -> DataFrame[RawDataSchema]:
```

Now this is independent of the Kedro catalog, but we could tie it back to Kedro world through introspection

**Comment by Galileo-Galilei** [comment]
Thanks @datajoely for the information about the ``check_types`` decorator. I think this is the "right way" to integrate pandera to standard function (mostly because when you don't have the dataaset/pipeline abstractions, you want to be protected against running the function against invalid data), but I think this is not necessarily what we should encourage to do. Conceptually, schema belongs to datasets and we should keep the decoupling between I/O and compute that kedro offers to have more flexibility (e.g. to turn validation off, to validate data manually and not at runtime, visualize data schema, document the pipeline...)

> Additionally - shall we make a little feature backlog, because we can split up the work?

Absolutely agree on this. I suggest to keep what you do in the kedro repo: 
- we can open issues for each feature request we have in mind, an describe what we want to work on. This enable doing things aysnchronously because I use to read issues during the day on my phone even if I don't have access to my latptop and can code. 
- if a feature is ill defined / too big, we can split it up in smaller ones and eventually keep it as a parent. We should not be afraid to close some issues once things are clearer to create smaller tickets.
- if some features appear to belong to a much broader topic, we can eventually crearte milestones to keep our ideas clear. 

At this stage (no users, not compatible with any public kedro version, no public declaration that the plugin exist, 8 stars), I am really comfortable with having a fast development pace, breaking things and iterating so feel free to go and merge without my validation. Havings an issue describing roughly what is at stake is enough for synchronization for now IMHO. 

@noklam I've just rebased so it may have messed up your hsitory. It's likely better to start from the main branch, i'll proceed on merging it

**Comment by noklam** [comment]
I just go on to create some issues first - I will fill that issues up eventually as I also have habit to read/write my issues during my commute without internet.

**Pull request merged**
